Source/Core/DolphinQt2/Settings/GameCubePane.cpp

```diff
@@ -0,0 +1,333 @@
+  {
+    m_slot_combos[2]->addItem(device);
+  }
+
+  device_layout->addWidget(new QLabel(tr("Slot A:")), 0, 0);
+  device_layout->addWidget(m_slot_combos[0], 0, 1);
+  device_layout->addWidget(m_slot_buttons[0], 0, 2);
+  device_layout->addWidget(new QLabel(tr("Slot B:")), 1, 0);
+  device_layout->addWidget(m_slot_combos[1], 1, 1);
+  device_layout->addWidget(m_slot_buttons[1], 1, 2);
+  device_layout->addWidget(new QLabel(tr("SP1:")), 2, 0);
+  device_layout->addWidget(m_slot_combos[2], 2, 1);
+
+  layout->addWidget(ipl_box);
+  layout->addWidget(device_box);
+
+  setLayout(layout);
+}
+
+void GameCubePane::ConnectWidgets()
+{
+  // IPL Settings
+  connect(m_skip_main_menu, &QCheckBox::stateChanged, this, &GameCubePane::SaveSettings);
+  connect(m_language_combo, static_cast<void (QComboBox::*)(int)>(&QComboBox::currentIndexChanged),
+          this, &GameCubePane::SaveSettings);
+  connect(m_override_language_ntsc, &QCheckBox::stateChanged, this, &GameCubePane::SaveSettings);
+
+  // Device Settings
+  for (int i = 0; i < 3; i++)
+  {
+    connect(m_slot_combos[i],
+            static_cast<void (QComboBox::*)(int)>(&QComboBox::currentIndexChanged), this,
+            &GameCubePane::SaveSettings);
+    if (i < 2)
+    {
+      connect(m_slot_buttons[i], &QPushButton::pressed, this, [this, i] { OnConfigPressed(i); });
+    }
+  }
+}
+
+void GameCubePane::OnConfigPressed(int slot)
+{
+  QString filter;
+  bool memcard = false;
+
+  switch (m_slot_combos[slot]->currentIndex())
+  {
+  // Memory card
+  case 2:
+    filter = tr("GameCube Memory Cards (*.raw,*.gcp)");
+    memcard = true;
+    break;
+  // Advance Game Port
+  case 5:
+    filter = tr("Game Boy Advance Carts (*.gba)");
+    memcard = false;
+    break;
+  // Microphone
+  case 6:
+    MappingWindow(this, MappingWindow::Type::MAPPING_GC_MICROPHONE, slot).exec();
+    return;
+  }
+
+  QString filename =
+      QFileDialog::getOpenFileName(this, tr("Choose a file to open"), QStringLiteral(""), filter);
+
+  if (filename.isEmpty() || !File::Exists(filename.toStdString()))
+    return;
+
+  QString path_abs = QFileInfo(filename).absolutePath();
+
+  // Memcard validity checks
+  if (memcard)
+  {
+    GCMemcard mc(filename.toStdString());
+
+    if (!mc.IsValid())
+    {
+      QMessageBox::critical(this, tr("Error"), tr("Cannot use that file as a memory card.\n%s\n"
+                                                  "is not a valid GameCube memory card file")
+                                                   .arg(filename));
+      return;
+    }
+
+    QString path_b =
+        QFileInfo(QString::fromStdString(slot == 0 ? SConfig::GetInstance().m_strMemoryCardB :
+                                                     SConfig::GetInstance().m_strMemoryCardA))
+            .absolutePath();
+
+    if (path_abs == path_b)
+    {
+      QMessageBox::critical(this, tr("Error"),
+                            tr("Are you trying to use the same file in both slots?"));
```

Comment:
The phrasing feels weird for an error message here. I'm not sure this should be written as a question :P

Reply:
I just copied it from Wx but I'd have no problem replacing it. Would "You can't use the same file in both slots!" be more appropriate?

Reply:
If we want to have a message like that, I would prefer the wording "The same file can't be used in both slots."

Reply:
Done.